File: init-service/app/services/cas_gc_service.py
import asyncio
import logging
from datetime import UTC, datetime, timedelta

from botocore.exceptions import BotoCoreError, ClientError, EndpointConnectionError
from sqlalchemy import select

from app.config import settings
from app.db import session_scope
from app.models.project import Node
from app.services.cas_service import CAS_PREFIX, s3_client


logger = logging.getLogger(__name__)
DELETE_BATCH_LIMIT = 1000


def _normalize_key_prefix(prefix: str) -> str:
    return prefix.rstrip("/") + "/"


def _extract_hash_from_key(key: str) -> str | None:
    prefix = _normalize_key_prefix(CAS_PREFIX)
    if not key.startswith(prefix):
        return None

    parts = key[len(prefix) :].split("/")
    if len(parts) != 2:
        return None

    shard, content_hash = parts
    if len(shard) != 2 or len(content_hash) != 64:
        return None
    if content_hash[:2] != shard:
        return None
    return content_hash


def get_live_content_hashes() -> set[str]:
    with session_scope() as session:
        rows = session.scalars(
            select(Node.content_hash).where(Node.content_hash.is_not(None)).distinct()
        ).all()
    return {content_hash for content_hash in rows if content_hash}


def collect_orphaned_objects(*, grace_hours: int) -> tuple[list[dict[str, object]], int]:
    bucket = settings.s3_bucket
    if not bucket:
        return [], 0

    live_hashes = get_live_content_hashes()
    prefix = _normalize_key_prefix(CAS_PREFIX)
    grace_cutoff = datetime.now(UTC) - timedelta(hours=grace_hours)
    paginator = s3_client.get_paginator("list_objects_v2")

    candidates: list[dict[str, object]] = []
    scanned = 0

    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
        for obj in page.get("Contents", []):
            key = obj.get("Key")
            if not key:
                continue
            scanned += 1

            content_hash = _extract_hash_from_key(key)
            if not content_hash or content_hash in live_hashes:
                continue

            last_modified = obj.get("LastModified")
            if last_modified is None or last_modified > grace_cutoff:
                continue

            candidates.append(
                {
                    "Key": key,
                    "hash": content_hash,
                    "last_modified": last_modified,
                    "size": obj.get("Size", 0),
                }
            )

    return candidates, scanned


def delete_orphaned_objects(orphaned_objects: list[dict[str, object]], *, batch_size: int) -> int:
    bucket = settings.s3_bucket
    if not bucket or not orphaned_objects:
        return 0

    normalized_batch_size = max(1, min(batch_size, DELETE_BATCH_LIMIT))
    deleted = 0

    for start in range(0, len(orphaned_objects), normalized_batch_size):
        chunk = orphaned_objects[start : start + normalized_batch_size]
        response = s3_client.delete_objects(
            Bucket=bucket,
            Delete={"Objects": [{"Key": str(item["Key"])} for item in chunk], "Quiet": True},
        )
        errors = response.get("Errors", [])
        if errors:
            raise RuntimeError(f"Failed to delete {len(errors)} CAS objects: {errors}")
        deleted += len(response.get("Deleted", []))

    return deleted


def run_cas_gc_once(*, grace_hours: int | None = None, dry_run: bool = False, batch_size: int | None = None) -> dict:
    bucket = settings.s3_bucket
    if not bucket:
        logger.info("CAS GC skipped because S3_BUCKET is not configured.")
        return {
            "status": "skipped",
            "reason": "missing_bucket",
            "scanned": 0,
            "candidates": 0,
            "deleted": 0,
            "dry_run": dry_run,
        }

    effective_grace_hours = grace_hours if grace_hours is not None else settings.cas_gc_grace_hours
    effective_batch_size = batch_size if batch_size is not None else settings.cas_gc_batch_size

    orphaned_objects, scanned = collect_orphaned_objects(grace_hours=effective_grace_hours)
    deleted = 0
    if not dry_run and orphaned_objects:
        deleted = delete_orphaned_objects(orphaned_objects, batch_size=effective_batch_size)

    result = {
        "status": "ok",
        "bucket": bucket,
        "prefix": _normalize_key_prefix(CAS_PREFIX),
        "scanned": scanned,
        "candidates": len(orphaned_objects),
        "deleted": deleted,
        "dry_run": dry_run,
        "grace_hours": effective_grace_hours,
    }
    logger.info("CAS GC completed: %s", result)
    return result


async def cas_gc_loop() -> None:
    interval_seconds = max(60, settings.cas_gc_interval_minutes * 60)
    while True:
        try:
            await asyncio.to_thread(run_cas_gc_once)
        except (BotoCoreError, ClientError, EndpointConnectionError, OSError, RuntimeError) as exc:
            logger.warning("CAS GC run failed: %s", exc)
        except Exception:
            logger.exception("CAS GC loop crashed unexpectedly.")
        await asyncio.sleep(interval_seconds)

File: init-service/src/services/cas_gc_service.ts
import { ListObjectsV2Command, DeleteObjectsCommand } from "@aws-sdk/client-s3";
import { s3Client } from "./s3_service";
import config from "../config";
import logger from "../loaders/logger";
import { db } from "../loaders/postgres";
import { nodes as nodesTable } from "db-schema";
import { isNotNull } from "drizzle-orm";

const CAS_PREFIX = "yuvro/CAS";
const DELETE_BATCH_LIMIT = 1000;

function normalizeKeyPrefix(prefix: string): string {
  return prefix.endsWith("/") ? prefix : prefix + "/";
}

function extractHashFromKey(key: string): string | null {
  const prefix = normalizeKeyPrefix(CAS_PREFIX);
  if (!key.startsWith(prefix)) return null;

  const rest = key.substring(prefix.length);
  const parts = rest.split("/");
  if (parts.length !== 2) return null;

  const [shard, contentHash] = parts;
  if (shard.length !== 2 || contentHash.length !== 64) return null;
  if (contentHash.substring(0, 2) !== shard) return null;

  return contentHash;
}

export async function getLiveContentHashes(): Promise<Set<string>> {
  const rows = await db
    .selectDistinct({ contentHash: nodesTable.contentHash })
    .from(nodesTable)
    .where(isNotNull(nodesTable.contentHash));

  const set = new Set<string>();
  for (const row of rows) {
    if (row.contentHash) {
      set.add(row.contentHash);
    }
  }
  return set;
}

export async function collectOrphanedObjects(graceHours: number): Promise<{
  candidates: Array<{ Key: string; hash: string; last_modified?: Date; size: number }>;
  scanned: number;
}> {
  const bucket = config.S3_BUCKET;
  if (!bucket) return { candidates: [], scanned: 0 };

  const liveHashes = await getLiveContentHashes();
  const prefix = normalizeKeyPrefix(CAS_PREFIX);
  const graceCutoff = new Date(Date.now() - graceHours * 60 * 60 * 1000);

  const candidates: Array<{ Key: string; hash: string; last_modified?: Date; size: number }> = [];
  let scanned = 0;
  let continuationToken: string | undefined;

  do {
    const listCommand: any = {
      Bucket: bucket,
      Prefix: prefix,
    };
    if (continuationToken) {
      listCommand.ContinuationToken = continuationToken;
    }

    const response = await s3Client.send(new ListObjectsV2Command(listCommand));
    const contents = response.Contents || [];

    for (const obj of contents) {
      const key = obj.Key;
      if (!key) continue;
      scanned++;

      const contentHash = extractHashFromKey(key);
      if (!contentHash || liveHashes.has(contentHash)) continue;

      const lastModified = obj.LastModified;
      if (!lastModified || lastModified > graceCutoff) continue;

      candidates.push({
        Key: key,
        hash: contentHash,
        last_modified: lastModified,
        size: obj.Size || 0,
      });
    }

    continuationToken = response.IsTruncated ? response.NextContinuationToken : undefined;
  } while (continuationToken);

  return { candidates, scanned };
}

export async function deleteOrphanedObjects(
  orphanedObjects: Array<{ Key: string }>,
  batchSize: number
): Promise<number> {
  const bucket = config.S3_BUCKET;
  if (!bucket || orphanedObjects.length === 0) return 0;

  const normalizedBatchSize = Math.max(1, Math.min(batchSize, DELETE_BATCH_LIMIT));
  let deletedCount = 0;

  for (let start = 0; start < orphanedObjects.length; start += normalizedBatchSize) {
    const chunk = orphanedObjects.slice(start, start + normalizedBatchSize);
    const response = await s3Client.send(
      new DeleteObjectsCommand({
        Bucket: bucket,
        Delete: {
          Objects: chunk.map((item) => ({ Key: item.Key })),
          Quiet: true,
        },
      })
    );

    const errors = response.Errors || [];
    if (errors.length > 0) {
      throw new Error(`Failed to delete ${errors.length} CAS objects.`);
    }

    deletedCount += (response.Deleted || []).length;
  }

  return deletedCount;
}

export async function runCasGcOnce(
  graceHours?: number,
  dryRun: boolean = false,
  batchSize?: number
): Promise<any> {
  const bucket = config.S3_BUCKET;
  if (!bucket) {
    logger.info("CAS GC skipped because S3_BUCKET is not configured.");
    return {
      status: "skipped",
      reason: "missing_bucket",
      scanned: 0,
      candidates: 0,
      deleted: 0,
      dry_run: dryRun,
    };
  }

  const effectiveGraceHours = graceHours !== undefined ? graceHours : config.CAS_GC_GRACE_HOURS;
  const effectiveBatchSize = batchSize !== undefined ? batchSize : config.CAS_GC_BATCH_SIZE;

  const { candidates, scanned } = await collectOrphanedObjects(effectiveGraceHours);
  let deleted = 0;

  if (!dryRun && candidates.length > 0) {
    deleted = await deleteOrphanedObjects(candidates, effectiveBatchSize);
  }

  const result = {
    status: "ok",
    bucket,
    prefix: normalizeKeyPrefix(CAS_PREFIX),
    scanned,
    candidates: candidates.length,
    deleted,
    dry_run: dryRun,
    grace_hours: effectiveGraceHours,
  };

  logger.info(`CAS GC completed: ${JSON.stringify(result)}`);
  return result;
}

export function startCasGcLoop(): void {
  const intervalMs = Math.max(60 * 1000, config.CAS_GC_INTERVAL_MINUTES * 60 * 1000);
  
  const loop = async () => {
    try {
      await runCasGcOnce();
    } catch (error) {
      logger.warn(`CAS GC run failed: ${error}`);
    } finally {
      setTimeout(loop, intervalMs);
    }
  };

  if (config.CAS_GC_ENABLED) {
    logger.info(`Starting CAS GC loop with interval: ${config.CAS_GC_INTERVAL_MINUTES} minutes`);
    setTimeout(loop, intervalMs);
  }
}
